src/puffin/src/puffin_manager/cached_puffin_manager/writer.rs

```diff
@@ -0,0 +1,193 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashSet;
+use std::path::PathBuf;
+
+use async_compression::futures::bufread::ZstdEncoder;
+use async_trait::async_trait;
+use futures::io::BufReader;
+use futures::{AsyncRead, AsyncSeek, AsyncWrite, StreamExt};
+use snafu::{ensure, ResultExt};
+use tokio_util::compat::TokioAsyncReadCompatExt;
+use uuid::Uuid;
+
+use crate::blob_metadata::CompressionCodec;
+use crate::error::{
+    DuplicateBlobSnafu, MetadataSnafu, OpenSnafu, Result, SerializeJsonSnafu,
+    UnsupportedCompressionSnafu, WalkDirSnafu,
+};
+use crate::file_format::writer::{Blob, PuffinAsyncWriter, PuffinFileWriter};
+use crate::puffin_manager::cache_manager::CacheManagerRef;
+use crate::puffin_manager::cached_puffin_manager::{DirFileMetadata, DirMetadata};
+use crate::puffin_manager::{PuffinWriter, PutOptions};
+
+/// `CachedPuffinWriter` is a `PuffinWriter` that writes blobs and directories to a puffin file.
+pub struct CachedPuffinWriter<CR, W> {
+    /// The name of the puffin file.
+    puffin_file_name: String,
+
+    /// The cache manager.
+    cache_manager: CacheManagerRef<CR>,
+
+    /// The underlying `PuffinFileWriter`.
+    puffin_file_writer: PuffinFileWriter<W>,
+
+    /// Written blob keys.
+    blob_keys: HashSet<String>,
+}
+
+#[async_trait]
+impl<CR, W> PuffinWriter for CachedPuffinWriter<CR, W>
+where
+    CR: AsyncRead + AsyncSeek,
+    W: AsyncWrite + Unpin + Send,
+{
+    async fn put_blob<R>(&mut self, key: &str, raw_data: R, options: PutOptions) -> Result<u64>
```

Comment:
It seems we should put blob into `cache_manager` too.

Reply:
The reason for the difference in behavior between `blob` and `dir` here is:

1. When the user prepares a `dir` for writing, `dir` can be repeatedly read, and it is assumed that the user has relinquished ownership of `dir`. At this point, the cache can take advantage of this by directly moving the `dir` to the cache directory.
2. However, for `blob`, the reader provided by the user is not necessarily a file in the filesystem. This reader cannot be repeatedly read. If it must be implemented, the process would involve writing to the cache first, then reading back from the cache to write into puffin. But handling the potential issues with interruptions during this process is cumbersome and undesirable.

Here, I will treat the handling of `dir` as a special case that takes full advantage of the user's `dir` characteristics, rather than deliberately aligning the behavior of handling `blob` with that of handling `dir`.